README.md
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/stactools-packages/worldpop/main?filepath=docs/installation_and_basic_usage.ipynb)

# stactools-worldpop

- Name: worldpop
- Package: `stactools.worldpop`
- PyPI: https://pypi.org/project/stactools-worldpop/
- Owner: @jamesvrt
- Dataset homepage: https://www.worldpop.org/
- STAC extensions used:
  - [proj](https://github.com/stac-extensions/projection/)
  - [scientific](https://github.com/stac-extensions/scientific/)
  - [raster](https://github.com/stac-extensions/raster/)
  - [item-assets](https://github.com/stac-extensions/item-assets/)

## Dataset Structure

WorldPop is a collection of 15 "projects" each containing several "categories" covering
 many demographic products and the supporting data used to derive them. This
 package is limited to two projects and two categories within each, totalling four
 categories:

- Project: **Population counts** (`pop`) [[link](https://www.worldpop.org/project/categories?id=3)]
  - Category: Unconstrained, UN adjusted, 100m (`wpgpunadj`) [[link](https://www.worldpop.org/geodata/listing?id=69)]
  - Category: Constrained, UN adjusted, 100m (`cic2020_UNadj_100m`) [[link](https://www.worldpop.org/geodata/listing?id=79)]
- Project: **Population age and sex structure** (`age_structures`) [[link](https://www.worldpop.org/project/categories?id=8)]
  - Category: Unconstrained, 100m (`aswpgp`) [[link](https://www.worldpop.org/geodata/listing?id=30)]
  - Category: Constrained, UN adjusted, 100m (`ascicua_2020`) [[link](https://www.worldpop.org/geodata/listing?id=88)]

You can read more about unconstrained vs. constrained methods [here](https://www.worldpop.org/methods/top_down_constrained_vs_unconstrained).

## STAC Structure

The STAC is structured to match the project/category format of WorldPop, resulting in
 four separate STAC Collections. Each item in each Collection covers a country.

- Collection: Age and sex pop. structures (unconstrained, 100m)
    - Items: Countries \* years
         - Assets: Ages \* sexes
- Collection: Age and sex pop. structures (constrained, UN adj., 100m)
    - Items: Countries \* 2020
         - Assets: Ages \* sexes
- Collection: Population counts (unconstrained, UN adj., 100m)
    - Items: Countries \* years
         - Assets: Ages \* sexes
- Collection: Population counts (constrained, UN adj., 100m)
    - Items: Countries \* 2020
         - Assets: Ages \* sexes

## STAC Examples

- [Collection](examples/collection.json)
- [Item](examples/FIN_2020.json)

## Command-line usage

To create one Item:

```bash
$ stac worldpop create-item -d destination
```

To create one Collection (not populated with Items):

```bash
$ stac worldpop create-collection -d destination
```

To create one Collection and populate it with Items:

```bash
$ stac worldpop populate-collection -d destination
```

To create all Collections and populate them with Items:

```bash
$ stac worldpop populate-all-collections -d destination
```

Use `stac worldpop --help` to see all subcommands and options.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [Unreleased]

### Added

- Updated to stactools 0.2.3 [#3](https://github.com/stactools-packages/worldpop/pull/3)
- Made the `mypy` configuration more strict [#3](https://github.com/stactools-packages/worldpop/pull/3)
- Included capabilities to download data assets, create COGs, tile COGs.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- CLI was failing due to improperly formatted help strings [#3](https://github.com/stactools-packages/worldpop/pull/3)


tests/test_stac.py
import unittest

from stactools.worldpop import stac
from stactools.worldpop.constants import API_URL
from stactools.worldpop.utils import get_metadata


class StacTest(unittest.TestCase):
    def test_create_collection(self):
        project = "pop"
        category = "wpgpunadj"
        collection = stac.create_collection(project, category)
        collection.set_self_href("")

        # Check that it has some required attributes
        self.assertEqual(collection.id, f"{project}_{category}")
        self.assertEqual(collection.license, "CC-BY-4.0")
        self.assertEqual(collection.extra_fields["sci:doi"],
                         "10.5258/SOTON/WP00660")
        self.assertEqual(len(collection.extra_fields["item_assets"]), 3)
        self.assertEqual(collection.summaries.lists["gsd"], [100.0])

        # Validate
        collection.validate()

    def test_create_item(self):
        project = "pop"
        category = "wpgpunadj"
        iso3 = "CHN"
        popyear = "2020"
        metadata_url = f"{API_URL}/{project}/{category}?iso3={iso3}"
        metadatas = get_metadata(metadata_url)["data"]
        item = stac.create_item(project, category, iso3, popyear, metadatas)

        # Check that it has some required attributes
        self.assertEqual(item.id, f"{iso3}_{popyear}")
        self.assertEqual(item.properties["gsd"], 100.0)
        self.assertEqual(item.properties["sci:doi"], '10.5258/SOTON/WP00660')
        self.assertEqual(item.properties["proj:epsg"], 4326)
        self.assertEqual(len(item.assets), 3)

        # Validate
        item.validate()


tests/test_module.py
import unittest

import stactools.worldpop


class TestModule(unittest.TestCase):
    def test_version(self):
        self.assertIsNotNone(stactools.worldpop.__version__)


tests/test_commands.py
import os.path
from tempfile import TemporaryDirectory

import pystac
from stactools.testing import CliTestCase

from stactools.worldpop.commands import create_worldpop_command


class CommandsTest(CliTestCase):
    def create_subcommand_functions(self):
        return [create_worldpop_command]

    def test_create_collection(self):
        with TemporaryDirectory() as tmp_dir:
            collection_dest = os.path.join(tmp_dir, "pop_wpgpunadj")

            result = self.run_command(
                ["worldpop", "create-collection", "-d", tmp_dir])

            self.assertEqual(result.exit_code,
                             0,
                             msg="\n{}".format(result.output))

            jsons = [
                p for p in os.listdir(collection_dest) if p.endswith(".json")
            ]
            self.assertEqual(len(jsons), 1)

            collection = pystac.read_file(
                os.path.join(collection_dest, jsons[0]))
            self.assertEqual(collection.id, "pop_wpgpunadj")
            self.assertEqual(collection.license, "CC-BY-4.0")
            self.assertEqual(collection.extra_fields["sci:doi"],
                             "10.5258/SOTON/WP00660")
            self.assertEqual(len(collection.extra_fields["item_assets"]), 3)
            self.assertEqual(collection.summaries.lists["gsd"], [100.0])

            collection.validate()

    def test_create_item(self):
        with TemporaryDirectory() as tmp_dir:
            result = self.run_command([
                "worldpop",
                "create-item",
                "-d",
                tmp_dir,
            ])
            self.assertEqual(result.exit_code,
                             0,
                             msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(tmp_dir) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            item = pystac.read_file(os.path.join(tmp_dir, jsons[0]))
            self.assertEqual(item.id, "CHN_2020")
            self.assertEqual(item.properties["gsd"], 100.0)
            self.assertEqual(item.properties["sci:doi"],
                             '10.5258/SOTON/WP00660')
            self.assertEqual(item.properties["proj:epsg"], 4326)
            self.assertEqual(len(item.assets), 3)

            item.validate()


src/stactools/worldpop/cog.py
import logging
import os
from glob import glob
from subprocess import CalledProcessError, check_output
from tempfile import TemporaryDirectory
from typing import Dict, List
from zipfile import ZipFile

import rasterio
import requests

from stactools.worldpop.constants import API_URL, TILING_PIXEL_SIZE
from stactools.worldpop.utils import get_iso3_list, get_metadata

logger = logging.getLogger(__name__)


def download_create_cog(
    output_directory: str,
    access_url: str,
    retile: bool = False,
    raise_on_fail: bool = True,
    dry_run: bool = False,
) -> str:
    if dry_run:
        logger.info("Would have downloaded TIFF, created COG, and written COG")
        return output_directory

    with TemporaryDirectory() as tmp_dir:
        # Extract filename from url
        tmp_file = os.path.join(tmp_dir, access_url.split("/").pop())

        logger.info("Downloading TIFF")
        logger.debug(f"access_url: {access_url}")
        resp = requests.get(access_url)

        with open(tmp_file, "wb") as f:
            logger.info("Writing TIFF")
            logger.debug(f"tmp_file: {tmp_file}")
            f.write(resp.content)
        if access_url.endswith(".zip"):
            logger.info("Unzipping TIFF")
            with ZipFile(tmp_file, "r") as zip_ref:
                zip_ref.extractall(tmp_dir)
        file_name = glob(f"{tmp_dir}/*.tif").pop()
        if retile:
            return create_retiled_cogs(file_name, output_directory,
                                       raise_on_fail, dry_run)
        else:
            output_file = os.path.join(
                output_directory,
                os.path.basename(file_name).replace(".tif", "") + "_cog.tif",
            )
            return create_cog(file_name, output_file, raise_on_fail, dry_run)


def create_retiled_cogs(
    input_path: str,
    output_directory: str,
    raise_on_fail: bool = True,
    dry_run: bool = False,
) -> str:
    """Split tiff into tiles and create COGs

    Args:
        input_path (str): Path to the input raster
        output_directory (str): The directory to which the COG will be written.
        raise_on_fail (bool, optional): Whether to raise error on failure.
            Defaults to True.
        dry_run (bool, optional): Run without downloading tif, creating COG,
            and writing COG. Defaults to False.

    Returns:
        str: The path to the output COGs.
    """
    output = None
    try:
        if dry_run:
            logger.info(
                "Would have split TIFF into tiles, created COGs, and written COGs"
            )
        else:
            logger.info("Retiling TIFF")
            logger.debug(f"input_path: {input_path}")
            logger.debug(f"output_directory: {output_directory}")
            with TemporaryDirectory() as tmp_dir:
                cmd = [
                    "gdal_retile.py",
                    "-ps",
                    str(TILING_PIXEL_SIZE[0]),
                    str(TILING_PIXEL_SIZE[1]),
                    "-targetDir",
                    tmp_dir,
                    input_path,
                ]
                try:
                    output = check_output(cmd)
                except CalledProcessError as e:
                    output = e.output
                    raise
                finally:
                    logger.info(f"output: {str(output)}")
                file_names = glob(f"{tmp_dir}/*.tif")
                for f in file_names:
                    input_file = os.path.join(tmp_dir, f)
                    output_file = os.path.join(
                        output_directory,
                        os.path.basename(f).replace(".tif", "") + "_cog.tif",
                    )
                    with rasterio.open(input_file, "r") as dataset:
                        contains_data = dataset.read().any()
                    # Exclude empty files
                    if contains_data:
                        logger.debug(f"Tile contains data: {input_file}")
                        create_cog(input_file, output_file, raise_on_fail,
                                   dry_run)
                    else:
                        logger.debug(f"Ignoring empty tile: {input_file}")

    except Exception:
        logger.error("Failed to process {}".format(input_path))

        if raise_on_fail:
            raise

    return output_directory


def create_cog(
    input_path: str,
    output_path: str,
    raise_on_fail: bool = True,
    dry_run: bool = False,
) -> str:
    """Create COG from a TIFF

    Args:
        input_path (str): Path to input raster
        output_path (str): The path to which the COG will be written.
        raise_on_fail (bool, optional): Whether to raise error on failure.
            Defaults to True.
        dry_run (bool, optional): Run without downloading TIFF, creating COG,
            and writing COG. Defaults to False.

    Returns:
        str: The path to the output COG.
    """

    output = None
    try:
        if dry_run:
            logger.info("Would have read TIFF, created COG, and written COG")
        else:
            logger.info("Converting TIFF to COG")
            logger.debug(f"input_path: {input_path}")
            logger.debug(f"output_path: {output_path}")
            cmd = [
                "gdal_translate",
                "-of",
                "COG",
                "-co",
                "NUM_THREADS=ALL_CPUS",
                "-co",
                "BLOCKSIZE=512",
                "-co",
                "COMPRESS=DEFLATE",
                "-co",
                "LEVEL=9",
                "-co",
                "PREDICTOR=YES",
                "-co",
                "OVERVIEWS=IGNORE_EXISTING",
                "-a_nodata",
                "0",
                input_path,
                output_path,
            ]

            try:
                output = check_output(cmd)
            except CalledProcessError as e:
                output = e.output
                raise
            finally:
                logger.info(f"output: {str(output)}")

    except Exception:
        logger.error("Failed to process {}".format(output_path))

        if raise_on_fail:
            raise

    return output_path


def tile_from_source(projects: Dict[str, List[str]], output_dir: str) -> None:
    """Created tiled cogs remote .tif files using a dict of project - categories pairs.

    Output files are named using the following convention:
    f"{project}_{category}_{iso}_{popyear}_{file_num}_{tile_x}_{tile_y}.tif"

    Args:
        projects (dict): Project - categories pairs in the form:
        ```
        {
            "project1": ["category1, "category2", ...],
            "project2": ["category1, "category2", ...],
            ...
        }
        ```
        output_dir (str): Output directory for tiles.
    """
    for project, categories in projects.items():
        for category in categories:
            for iso in get_iso3_list(project, category):

                metadata = get_metadata(
                    f"{API_URL}/{project}/{category}?iso3={iso}")["data"]
                years_and_files = {
                    data["popyear"]: data["files"]
                    for data in metadata
                }

                for popyear, files in years_and_files.items():
                    for file_num, tif in enumerate(files):

                        local_tif_name = (
                            f"{project}_{category}_{iso}_{popyear}_{file_num}.tif"
                        )

                        with TemporaryDirectory() as tmp_dir:
                            print(f"Downloading and tiling {tif}")
                            local_tif_path = os.path.join(
                                tmp_dir, local_tif_name)
                            with open(local_tif_path, "wb") as f:
                                response = requests.get(tif)
                                f.write(response.content)

                            create_retiled_cogs(local_tif_path, output_dir)


src/stactools/worldpop/utils.py
import json
import os
from datetime import datetime
from typing import Any, List
from urllib.parse import urlparse

import requests
from pystac.collection import Collection

from stactools.worldpop.constants import API_URL


def get_metadata(url: str) -> Any:
    """Return dictionary from JSON file at given path."""
    scheme = urlparse(url).scheme
    if scheme == "http" or scheme == "https":
        response = requests.get(url)
        if response.status_code != 200:
            raise AssertionError(f"API URL not found: {url}")
        return response.json()
    else:
        if not os.path.exists(url):
            raise AssertionError(f"File path not found: {url}")
        with open(url) as f:
            return json.load(f)


def get_iso3_list(project: str, category: str) -> List[str]:
    """Return a list of ISO3 country codes contained in a dataset/subset."""
    url = f"{API_URL}/{project}/{category}"
    response = requests.get(url)
    if response.status_code != 200:
        raise AssertionError(f"{response.status_code} code from API: {url}")
    data = response.json()["data"]
    return list(sorted(set([d["iso3"] for d in data])))


def get_popyears(collection: Collection) -> List[str]:
    """Return a list of years making up the temporale extent of a collection."""
    startstop: Any = collection.extent.temporal.intervals[0]
    start, stop_ = startstop
    if start is None:
        raise AssertionError("Collection's temporal extent starts at None")
    stop = stop_.year if stop_ is not None else datetime.now().year - 1
    return [str(y) for y in range(int(start.year), int(stop) + 1)]


src/stactools/worldpop/constants.py
# flake8: noqa

from typing import Any, Dict

from pyproj import CRS
from pystac import Link, Provider, ProviderRole

# TODO
WORLDPOP_ID = "worldpop"
WORLDPOP_EPSG = 4326
WORLDPOP_CRS = CRS.from_epsg(WORLDPOP_EPSG)
WORLDPOP_EXTENT = [-180., 90., 180., -90.]
WORLDPOP_TITLE = "WorldPop Country Datasets"
LICENSE = "CC-BY-4.0"
LICENSE_LINK = Link(
    rel="license",
    target="https://www.worldpop.org/data/licence.txt",
    title="Attribution 4.0 International (CC BY 4.0)",
)

WORLDPOP_DESCRIPTION = """WorldPop was initiated in October 2013 to combine the AfriPop, AsiaPop and AmeriPop population mapping projects. It aims to provide an open access archive of spatial demographic datasets for Central and South America, Africa and Asia to support development, disaster response and health applications."""

PROVIDERS = [
    Provider(name="WorldPop",
             roles=[
                 ProviderRole.HOST,
                 ProviderRole.LICENSOR,
                 ProviderRole.PROCESSOR,
                 ProviderRole.PRODUCER,
             ],
             url="https://www.worldpop.org/"),
    Provider(name="Bill and Melinda Gates Foundation",
             url="http://www.gatesfoundation.org/"),
    Provider(name="USAID", url="https://www.usaid.gov/"),
    Provider(name="UN Foundation", url="http://www.unfoundation.org/")
]

KEYWORDS = [
    "Population dynamics",
    "Population distributions",
    "Low income countries",
    "Middle income countries",
    "Spatial demographics",
    "Age and sex structures",
    "Population count",
    "Population",
    "Demographics",
    "Geographical maps",
]

API_URL = "https://www.worldpop.org/rest/data"

DESCRIPTION_POP = """<b>WorldPop produces different types of gridded population count datasets, depending on the methods used and end application. \r\nPlease make sure you have read our <a href=\"/methods/populations\" target=\"_blank\">Mapping Populations</a> overview page before choosing and downloading a dataset.</b> \r\n<br/><br/>   \r\nBespoke methods used to produce datasets for specific individual countries are available through the <b>WorldPop Open Population Repository (WOPR)</b> link below. \r\nThese are 100m resolution gridded population estimates using customized methods (\"<a href=\"https://www.worldpop.org/methods/populations\" target=\"_blank\">bottom-up</a>\" and/or \"<a href=\"https://www.worldpop.org/methods/populations\" target=\"_blank\">top-down</a>\") developed for the latest data available from each country. \r\nThey can also be visualised and explored through the <a href=\"https://apps.worldpop.org/woprVision/\" target=\"_blank\">woprVision App</a>.\r\n<br/>\r\nThe remaining datasets in the links below are produced using the \"<a href=\"/methods/populations\" target=\"_blank\">top-down</a>\" method, \r\nwith either the unconstrained or constrained top-down disaggregation method used. \r\nPlease make sure you read the <a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">Top-down estimation modelling overview page</a> to decide on which datasets best meet your needs. \r\nDatasets are available to download in Geotiff and ASCII XYZ format at a <u>resolution of 3 and 30 arc-seconds (approximately 100m and 1km at the equator, respectively)</u>:\r\n<br/><br/>\r\n- Unconstrained individual countries 2000-2020  ( 1km resolution ): Consistent 1km resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">unconstrained top-down methods</a> for all countries of the World for each year 2000-2020. \r\n<br/>\r\n- Unconstrained individual countries 2000-2020 ( 100m resolution ): Consistent 100m resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">unconstrained top-down methods</a> for all countries of the World for each year 2000-2020. \r\n<br/>\r\n- Unconstrained individual countries 2000-2020 UN adjusted ( 100m resolution ): Consistent 100m resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">unconstrained top-down methods</a> for all countries of the World for each year 2000-2020 and adjusted to match United Nations national population estimates (<a href=\"https://population.un.org/wpp/Download/Files/1_Indicators (Standard)/EXCEL_FILES/1_Population/WPP2019_POP_F01_1_TOTAL_POPULATION_BOTH_SEXES.xlsx\" target=\"_blank\">UN 2019</a>)\r\n<br/>\r\n-Unconstrained individual countries 2000-2020 UN adjusted ( 1km resolution ): Consistent 1km resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">unconstrained top-down methods</a> for all countries of the World for each year 2000-2020 and adjusted to match United Nations national population estimates (<a href=\"https://population.un.org/wpp/Download/Files/1_Indicators (Standard)/EXCEL_FILES/1_Population/WPP2019_POP_F01_1_TOTAL_POPULATION_BOTH_SEXES.xlsx\" target=\"_blank\">UN 2019</a>).\r\n<br/>\r\n-Unconstrained global mosaics 2000-2020 ( 1km resolution ): Mosaiced 1km resolution versions of the \"Unconstrained individual countries 2000-2020\" datasets.\r\n<br/>\r\n-Constrained individual countries 2020 ( 100m resolution ): Consistent 100m resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">constrained top-down methods</a> for all countries of the World for 2020.\r\n<br/>\r\n-Constrained individual countries 2020 UN adjusted ( 100m resolution ): Consistent 100m resolution population count datasets created using \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">constrained top-down methods</a> for all countries of the World for 2020 and adjusted to match United Nations national \r\npopulation estimates (<a href=\"https://population.un.org/wpp/Download/Files/1_Indicators (Standard)/EXCEL_FILES/1_Population/WPP2019_POP_F01_1_TOTAL_POPULATION_BOTH_SEXES.xlsx\" target=\"_blank\">UN 2019</a>).\r\n<br/>\r\n<br/>\r\nOlder datasets produced for specific individual countries and continents, using a set of tailored geospatial inputs and differing \"top-down\" methods and time periods are still available for download here: <a href=\"https://www.worldpop.org/geodata/listing?id=16\">Individual countries</a> and <a href=\"https://www.worldpop.org/geodata/listing?id=17\">Whole Continent</a>.\r\n"""
DESCRIPTION_AGE_SEX = """<b>WorldPop produces different types of gridded population count datasets, depending on the methods used and end application. \r\nPlease make sure you have read our <a href=\"/methods/populations\" target=\"_blank\">Mapping Populations</a> overview page before choosing and downloading a dataset.</b> \r\n<br/><br/>   \r\n\r\nA description of the modelling methods used for age and sex structures can be found in \r\n<a href=\"https://pophealthmetrics.biomedcentral.com/articles/10.1186/1478-7954-11-11\" target=\"_blank\">\r\nTatem et al</a> and \r\n<a href=\"https://dx.doi.org/10.1038/sdata.2017.89\" target=\"_blank\">Pezzulo et al</a>. Details of the input population count datasets used can be found <a href=\"https://www.worldpop.org/project/categories?id=3\" target=\"_blank\">here</a>, and age/sex structure proportion datasets <a href=\"https://www.portal.worldpop.org/demographics/data/global_agesex_proportions_totals_2020.zip\" target=\"_blank\">here</a>. \r\n<br/>\r\nBoth top-down 'unconstrained' and 'constrained' versions of the datasets are available, and the differences between the two methods are outlined \r\n<a href=\"/methods/top_down_constrained_vs_unconstrained\" target=\"_blank\">here</a>. The  datasets represent the outputs from a project focused on construction of consistent 100m resolution population count datasets for all countries of the World structured by male/female and 5-year age classes (plus a <1 year class). These efforts necessarily involved some shortcuts for consistency. The unconstrained datasets are available for each year from 2000 to 2020.  \r\nThe constrained datasets are only available for 2020 at present, given the time periods represented by the building footprint and built settlement datasets used in the mapping."""
TIME_EXTENT_2020 = ["2020-01-01T00:00:00Z", "2020-12-31T00:00:00Z"]
TIME_EXTENT_2000_ = ["2000-01-01T00:00:00Z", None]

COLLECTIONS_METADATA: Dict[str, Any] = {
    "pop": {
        "wpgpunadj": {
            "id":
            "pop_wpgpunadj",
            "title":
            """Population counts, unconstrained individual countries 2000-2020 UN adjusted (100m resolution)""",
            "description":
            DESCRIPTION_POP,
            "time_extent":
            TIME_EXTENT_2000_,
            "gsd":
            100.,
            "doi":
            "10.5258/SOTON/WP00660",
            "citation":
            """WorldPop (www.worldpop.org - School of Geography and Environmental Science, University of Southampton; Department of Geography and Geosciences, University of Louisville; Departement de Geographie, Universite de Namur) and Center for International Earth Science Information Network (CIESIN), Columbia University (2018). Global High Resolution Population Denominators Project - Funded by The Bill and Melinda Gates Foundation (OPP1134076). https://dx.doi.org/10.5258/SOTON/WP00660"""
        },
        "cic2020_UNadj_100m": {
            "id":
            "pop_cic2020_UNadj_100m",
            "title":
            """Population counts, constrained Individual countries 2020 UN adjusted (100m resolution)""",
            "description":
            DESCRIPTION_POP,
            "time_extent":
            TIME_EXTENT_2020,
            "gsd":
            100.,
            "doi":
            "10.5258/SOTON/WP00685",
            "citation":
            """Bondarenko M., Kerr D., Sorichetta A., and Tatem, A.J. 2020. Census/projection-disaggregated gridded population datasets, adjusted to match the corresponding UNPD 2020 estimates, for 183 countries in 2020 using Built-Settlement Growth Model (BSGM) outputs. WorldPop, University of Southampton, UK. doi:10.5258/SOTON/WP00685"""
        },
    },
    "age_structures": {
        "aswpgp": {
            "id":
            "age_structures_aswpgp",
            "title":
            """Age and sex structures, unconstrained individual countries 2000-2020 (100m resolution)""",
            "description":
            DESCRIPTION_AGE_SEX,
            "time_extent":
            TIME_EXTENT_2000_,
            "gsd":
            100.,
            "doi":
            "10.5258/SOTON/WP00646",
            "citation":
            """WorldPop (www.worldpop.org - School of Geography and Environmental Science, University of Southampton; Department of Geography and Geosciences, University of Louisville; Departement de Geographie, Universite de Namur) and Center for International Earth Science Information Network (CIESIN), Columbia University (2018). Global High Resolution Population Denominators Project - Funded by The Bill and Melinda Gates Foundation (OPP1134076). https://dx.doi.org/10.5258/SOTON/WP00646"""
        },
        "ascicua_2020": {
            "id":
            "age_structures_ascicua_2020",
            "title":
            """Age and sex structures, constrained individual countries 2020 UN adjusted (100m resolution)""",
            "description":
            DESCRIPTION_AGE_SEX,
            "time_extent":
            TIME_EXTENT_2020,
            "gsd":
            100.,
            "doi":
            "10.5258/SOTON/WP00698",
            "citation":
            """Bondarenko M., Kerr D., Sorichetta A., and Tatem, A.J. 2020. Estimates of 2020 total number of people per grid square, adjusted to match the corresponding UNPD 2020 estimates and broken down by gender and age groupings, produced using Built-Settlement Growth Model (BSGM) outputs. WorldPop, University of Southampton, UK. doi:10.5258/SOTON/WP00698"""
        },
    }
}

TILING_PIXEL_SIZE = (10001, 10001)
